### PR TITLE
Initial draft of Python PIP builder design doc

### PR DESCRIPTION
**Pull request by stealthycoin**
Initial draft of Python PIP builder design doc. Let me know what you think.

### CODE REVIEW

File: lambda_builders/actions/python_pip/DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+```python
+def build_depenencies(source_path, config=None):
+    """Builds a source directory's dependencies into the directory.
+
+	:type source_path: str
+	:param source_path: The source directory that needs to have dependencies
+		built for it. It is assumed that this customer code is correct and
+		would function if it had all of its dependencies installed alongside
+		it. It is assumed that the dependencies are specified in a top level
+		``requirements.txt`` file. This file can be removed before deploying
```
Comment (sanathkr) [inline lambda_builders/actions/python_pip/DESIGN.md]: I would assume that this method is passed a path to `requirements.txt`. SAM CLI will ensure the file is at top-level within the source folder but this should not be an assumption that low level build actions must make.

File: lambda_builders/actions/python_pip/DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		to lambda, but the this function will not modify any existing files
+		in this directory.
+
+	:type config: :class:`lambda_builders.actions.python_pip.config`
+	:param config: A configuration object which contains additional config
```
Comment (sanathkr) [inline lambda_builders/actions/python_pip/DESIGN.md]: I would call this out as optional, and not implemented, for now. This requirement exists for several build actions. So I want to standardize this as Optimizations that can turned on/off by customers - https://github.com/awslabs/aws-lambda-builders/issues/3

File: lambda_builders/actions/python_pip/DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+There are a couple ways error reporting can be done.
+
+1) We can raise a python error and have the caller catch and deal with
+   it at the end. The error may not be significant, for example if the user
+   is already aware of the problem and has a post-build hook that fixes the
+   bundle with a custom dependency injection. So this exception would be thrown
+   at the end and not interrupt the rest of the dependency building.
+2) Return None or a list of aggregated error objects that can be inspected by
+   the caller.
+3) Use the UI object. The builder as it is written right now has a notion of a
```
Comment (sanathkr) [inline lambda_builders/actions/python_pip/DESIGN.md]: I like the idea of a UI object. This allows us to abstract functionality like "write with colors" etc within the object. We could also redirect the data to a file transparently. Feel free to make assumptions about this object for now. We can implement a real UI object separately.

File: lambda_builders/actions/python_pip/DESIGN.md (hunk 2 above)
Comment (stealthycoin) [reply lambda_builders/actions/python_pip/DESIGN.md]: Ah yes I meant to write that, my bad.

File: lambda_builders/actions/python_pip/DESIGN.md (hunk 3 above)
Comment (stealthycoin) [reply lambda_builders/actions/python_pip/DESIGN.md]: Cool. For now Whatever wrapper around this low level builder can create and inject it.

File: lambda_builders/actions/python_pip/DESIGN.md (hunk 1 above)
Comment (stealthycoin) [reply lambda_builders/actions/python_pip/DESIGN.md]: Ok I can make this more in line with the params outlined in your other PR. I will still assume I have an adapter between them to make this low level builder more extractable and specific to python.

Comment (sanathkr) [reply lambda_builders/actions/python_pip/DESIGN.md]: Makes sense

File: lambda_builders/actions/python_pip/DESIGN.md (hunk 2 above)
Comment (sanathkr) [reply lambda_builders/actions/python_pip/DESIGN.md]: Ah I missed it ;)

**Comment by sanathkr** [comment]
Let me know once you have pushed a new draft. I will approve and merge

**Comment by sanathkr** [comment]
Accidentally deleted this file from the repo after creating the new directory structure. I will add this to my other open PR

**Pull request merged**